src/main/kotlin/com/exactpro/th2/codec/util/MessageUtil.kt

```diff
@@ -39,4 +50,45 @@ val MessageGroup.messageIds: List<MessageID>
             RAW_MESSAGE -> message.rawMessage.metadata.id
             else -> error("Unknown message kind: $kind")
         }
-    }
\ No newline at end of file
+    }
+
+fun MessageGroup.toErrorMessageGroup(exception: Throwable, protocol: String) : MessageGroup {
+    val result = MessageGroup.newBuilder()
+
+    val content = buildString {
+        appendLine("$protocol codec has failed to decode one of the following messages: ${messageIds.joinToString(", ") { it.toDebugString() }}")
+        appendLine("Due to the following errors: ")
+
+        generateSequence(exception, Throwable::cause).forEachIndexed { index, cause ->
+            appendLine("$index: ${cause.message}")
+        }
+    }
+
+    this.messagesList.forEach { message ->
+        when {
+            message.hasMessage() -> result.addMessages(message)
+            message.hasRawMessage() -> {
+                message.rawMessage.let { rawMessage ->
+                    if (rawMessage.metadata.protocol.run { isBlank() || this == protocol }) {
+                        result += message().apply {
+                            parentEventId = rawMessage.parentEventId
```

Comment:
You should set the `parentEventId` only if the original message `hasParentEventId`. Otherwise, you will take a default value and set it to the created message. It might trigger some unwanted behavior in other components what look for parentEventId to attach additional events to it

Reply:
As far I understand the worst thing that could happen in this case if we won't check that `rawMessage` has parent event id is that we'll replace default instance of `EventID` with the same default instance. How it's gonna cause an unwanted behavior?

Reply:
I think I get it - it would cause `hasParentEventId` to return `true`. Too bad that proto doesn't also check for default instances in such methods...

Reply:
The other boxes will check if the message has a parent ID by calling the `hasParentEventId()` method. If we set the parent ID here (even with a default value) the other side will get `true` as a result of the `hasParentEventId()` method. It will try to submit an event as a child to the empty EventID and this event does not exist. And we will get an event that does not have a parent